Add dist-upgrade and extra_args to zypper module

tsdmgz opened the pull request:
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
zypper

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0 (zypper-dup-function 04ad35b007) last updated 2017/02/12 23:34:35 (GMT +800)
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

Adds support for `zypper dist-upgrade` and `--no-allow-vendor-change`. This combination is a recommended method of keeping an openSUSE Tumbleweed system with additional repositories updated. 

Syntax follows semantics as set by `name='*' state=latest type=patch`, ideally invoked as `name='*' state=latest type=dist-upgrade no_allow_vendor_change=yes`.

robinro commented on lib/ansible/modules/packaging/os/zypper.py:
```diff
@@ -104,6 +104,13 @@
         required: false
         default: "no"
         choices: [ "yes", "no" ]
+    no_allow_vendor_change
```
missing `:` here.

robinro commented:
@tsdmgz 
Thanks for helping to improve the zypper module!

For `no-allow-vendor-change`: Can we avoid the double negative and put `allow-vendor-change` with default `yes`?
We could also instead add an option `extra_args`, which one can use to pass arbitrary arguments. There is also `allow-arch-change`, `allow-downgrade`, `allow-name-change` and `extra_args` would safe us from reimplementing the complete zypper API.

I can also see `dist-upgrade` integrated in the module. I don't think` type` is the right place to put it, though. For `patch` it makes more sense, since you can install individual patches that way, but for dist-upgrade that does not exist.

I'd prefer `state='dist-upgrade'`, with no other arguments or with `name='*'`. What do you think about that?
Another option would be to introduce an `upgrade` argument like in the `apt` module, but apt has more options there then zypper, so I'd rather use `state`.

tsdmgz commented:
On 西元2017年02月13日 01:53, Robin Roth wrote:
> Thanks for helping to improve the zypper module!

Much appreciated!

> For |no-allow-vendor-change|: Can we avoid the double negative and put |allow-vendor-change| with default |yes|?

Tried that at first, but couldn't exactly figure out how to make it work. Admittedly, that wording is confusing. Sorry, I'm really new with this and with Python in general and mostly followed what was written.

In hindsight: oh. wow. I can see how now. /headdesk

> We could also instead add an option |extra_args|, which one can use to pass arbitrary arguments. There is also |allow-arch-change|, |allow-downgrade|, |allow-name-change| and |extra_args| would safe us from reimplementing the complete zypper API.

Ah! Yes! `extra_args` is a better solution and gives us room to use this for other commands too. Should it be as a list or have it as a single line?

> I can also see |dist-upgrade| integrated in the module. I don't think|type| is the right place to put it, though. For |patch| it makes more sense, since you can install individual patches that way, but for dist-upgrade that does not exist.
> I'd prefer |state='dist-upgrade'|, with no other arguments or with |name='*'|. What do you think about that?
> Another option would be to introduce an |upgrade| argument like in the |apt| module, but apt has more options there then zypper, so I'd rather use |state|.

That's true, `dist-upgrade` isn't a package type and setting `state='absent'` at the same time would not be very useful. I'm for keeping `name='*'`. I don't know what the consequences of not making `name` required anymore, and follows how `zypper upgrade` is done.

robinro commented:
Sounds good.

Can you implement the `extra_args` argument (instead of allow-vendor-change)? I would use a string, since this is less ambigious. It should be added at the end of the command to allow overwriting of other arguments.

To remove `name=*` one has to add a more complex definition of the required argumens. It's possible but I agree keeping name required is easier and still intuitive.

tsdmgz commented:
Will do!

tsdmgz committed:
Implement state='dist-upgrade'

Implements `zypper dist-upgrade` for the zypper module. This follows
how `zypper upgrade` is invoked, except `state='dist-upgrade'`.

Setting name to anything other than '*' would cause the module to error
out. `dist-upgrade` affects all packages and would not make sense to
apply to a specific package.

tsdmgz committed:
Implement option extra_args

Add option to append additional arguments to zypper command. This
should be able to accommodate other options that are not (yet) covered
by zypper module.

Arguments are given as if written in the command line, complete with
dashes.

tsdmgz commented:
Sorry it took a while. I've been using it on my own setup and it looks okay. zypper does fail when it asks what to do with dependency problems, but that's because of --no-interactive I guess.

Question: does state=dist-upgrade fit in Ansible's declarative language? I tried to make it look that way in documentation but I'm not sure if it fits?

robinro commented on lib/ansible/modules/packaging/os/zypper.py:
```diff
@@ -57,9 +57,12 @@
-        choices: [ present, latest, absent ]
+        choices: [ present, latest, absent, dist-upgrade ]
         default: "present"
+        version_added: "2.3"
```
The `version_added` is not correct here, as it refers to `state` and not `dist-upgrade`. One could refer to the version in the doc string above, but I'd leave it out since it's an addition and the error message in old versions is descriptive enough.

robinro reviewed (changes requested):
Up to the minor points above this looks good to me. Once these are adjusted this can be merged. Thanks for including the changes!

One could check for inconsistencies more thoroughly (like dist-update together with type or force), but I guess the current state is a very appropriate compromise between covering typical mistakes and having readable code.

robinro commented on lib/ansible/modules/packaging/os/zypper.py:
```diff
@@ -304,6 +319,10 @@ def get_cmd(m, subcommand):
+    if m.params['extra_args']:
+        args_list = m.params['extra_args'].split(' ')
+        for extra_arg in args_list:
+            cmd.append(extra_arg)
```
Instead of for/append, use `cmd.extend(args_list)`

tsdmgz commented:
Done! Weirdly enough, tried using extend before but didn't work out and this time it did. Must've missed something.

Looking at it, checking for inconsistencies with dist-upgrade might involve breaking out the checks to its own function to keep it readable. But that could be wrong...

robinro commented:
shipit
Thanks @tsdmgz for this contribution!

Let's leave the checks the way they are and adjust them in the future as needed. I agree that their complexity could warrant a separate function.

tsdmgz commented:
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

Will try to take a crack at that some time in another PR 🎉
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEEEM6RUUr+JAwr+woLzFn16zCdzW0FAlisglIACgkQzFn16zCd
zW1cdg//VoMxoolo801itV401FwS0vQfzxws4CQQXxj5/deTw1bgqv7oqWGtRgf2
BcfflQ5/pnJJIH1nDv4WXuESqvqp0/01oFrAlUhrqmw47uxarLlBE2b7etY5+tBo
sQdeMM/kW6E1wZ+viszpdkCI22io/+fWW+JoEP+pe2bErbFv5onAZNxi30d3oYrW
gXBfc+UYnV31uIRaXJjldkcc9V/ccuFSgs7358NERvtFXDrVW20kayrBue2SCmMK
EfD87/2aabFtbQ5EsxjFlxuGKVslC02DlcV+XCxTaxEUybLcOJH6NYraToPoODZQ
xt+6yiuL0WTe+8k3xEYF1U69+TjrpJ8Zb/immz969YlJE91m3WsnPzgOjgk0CLZJ
eIeO6wwGThSzczPTBL7xQ9VSw/KwOfZlzanVgE4dgNb6pxHgbnLXAg/jFBFGOmK9
IzQl7OaSp1cCgtTmI31GA3y7oRoARuJtYvwKGuPzR+pfoW6jAJ/Sykpx8k3JXQ6W
tN8gkAK8vdCMS+ocCnnDQIwaa/bEaNeiQS7+5vdr8CaKoW9RdDK2nkfvunyp6bNC
AS+qexFklJoy1LKNfvxnzBQQKOGh6/79QEFrnvPKLnFTN/8H5bL1B7mAF54WRKyK
zwxMdJHgbsN6JuHoTkQWFZvteMT3W4EaSP3ogn++Z2lsbymcUn4=
=vvm8
-----END PGP SIGNATURE-----

alxgu commented on lib/ansible/modules/packaging/os/zypper.py:
```diff
@@ -108,6 +110,12 @@
         default: "no"
         choices: [ "yes", "no" ]
+    extra_args:
+        version_added: "2.3"
```
Please change it to "2.4"

tsdmgz commented:
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

sorry it took a while, slipped my mind. is this good?
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEEEM6RUUr+JAwr+woLzFn16zCdzW0FAlkceDQACgkQzFn16zCd
zW0YMRAAnychIvkfUicycp0wWdD77k43y1pcXVlkXBd0y8oCsvb6gQ8/HPjcEi5U
wUdw8cuf3NuE59fr+Izbrg6UeIhNjuCOLKFbDUzLuo9KajZyMJJaAOXvuby0G/b6
H4/nca+xC0yPpZxPmRCMKeEQSw0G4FCfZIYO3AdnuWDonEskqluOzLhv1yT7MtuM
OUKrD/KJ4j/2RFyuN2a2bDhutc7EtFgfzm8TJBFPqDv/XPADJFQJMLbgwp6PMbWE
/x8RHjJ6KMnODjdtckxV3pay7Rg8ZETbwJzcX/1RbTShEaGn2P5WC1Ri4jCP6n5l
Gb/BkaRxUTMovpsYNVK400wvKvFiZjAwRHWJqqWjV6Xytb/0woImMBnjrdL2drKe
RKBiDZTCmlIg2EJS7E/DhNVkm16vnhgakVR5ZTypf5+3BqzUcj4YQ/1D+VoJq1Jp
G2X2PoGXk3Y4IvfNe5JhXgDfYZcV+bQB/k54F9EOVu2otF7L9GivLhBg9PEedixf
IRAetUv6bQDtAYaQsDtXn6RpLcNIe/0rZgnJgLrt+I8Hu2GNULdxbEhz87/5mDrL
dmBXe2uzLawkZEXkKWIVJg1WtoiTF/1eF2clef9XJIStKdF7auNWdz6KWca88O13
Tsd40vdV+yE8+qFJ+LHz4Kyh7s7sVevlcqVoXpPZTHN0bo17OHI=
=w2AM
-----END PGP SIGNATURE-----

alxgu commented:
shipit

AnderEnder commented:
shipit

andytom commented:
shipit

pilou- commented:
It seems the `Request changes` review prevents the merge of this pull request: the bot added `needs_revision` label due to this `Request changes` review.

The bot will remove the `needs_revision` label if:
- either @robinro updates the Github review (I guess that's ok because @robinro commented using `shipit`)
- or @tsdmgz adds a comment containing `read_for_review`.

jctanner commented:
@robinro can you update your review on this?

robinro reviewed (approved):
shipit

robinro commented:
Sorry, I wasn't aware this is waiting on my review.
shipit

tsdmgz commented:
Thank you! Thought I did something wrong there ._.
